### DISCUSSION TITLE
concentration constraints

### DISCUSSION THREAD

**Discussion by dmebane** [CederGroupHub/smol#435] [Q&A] [answered]
Hi

I have created a cluster expansion and ensemble with four sublattices. Two of the sublattices are active. I would like to fix the concentration on one of the sublattices and fix the chemical potential on the other in a SGCMC simulation. I see in the example notebook how to create a CompositionSpace object that contains such constraints, but it's unclear how to pass that object to the sampler. Trying to parse the code but it's slow-going -- any help is appreciated!

**Comment by lbluque** [comment]
Hello @dmebane :wave: 

If I understand correctly, I think that your situation can be resolved without relying on the `CompositionSpace` class. Once you have created the `Ensemble`, you can set the sublattices you want to "freeze" from the two active sublattices to inactive by restricting all it's sites. Then make sure that the initial occupancy you use to start your MC sets the sites of that sublattice to the concentration you need. 

This will then result in MC where only sites of the single active sublattices (for which you are setting the chemical potential) will be updated. so the sites of those sublattices will never be changed, however the configuration of the originally active sublattice that you set to be inactive will still be used in computing the CE energy.

This would be something like this:
```python

assert len(ensemble.active_sublattices) == 2
assert len(ensemble.inactive_sublattices) == 2

# say the sublattice you want to set to inactive is index 0
ensemble.active_sublattices[0].restrict_sites(ensemble.active_sublattices[0].sites)

# This should render that sublattice as inactive now
assert len(ensemble.active_sublattices) == 1
assert assert len(ensemble.inactive_sublattices) == 3
```

Does this sound like it addresses your use case?

**Reply by dmebane** [reply]
Hi @lbluque thanks so much for your response! I considered restricting sites as you suggest, but I was concerned that would preclude moves on that sublattice, which still has partial occupancy. However if that’s not the case then yes, this addresses our issue . . . .

**Reply by lbluque** [reply]
Ah, I misunderstood that part. So the idea is that you want to have one of the two active sublattices to have a set chemical potential and the other to have fixed composition but still allow transitions/swaps at the composition, correct?

You're right that freezing all sites won't cut it if that's the case. I think this can be addressed by setting up table flip transitions with composition constraints as you initially suggested. @qchempku2017 may be able to help better there. (Though he is traveling currently).

A question though, are there any overlapping species in the two active sublattices, ie a species present in both?

**Reply by dmebane** [reply]
Yes, exactly -- ideally one sublattice has a fixed chemical potential and the other has a fixed composition. There is no mixing between the two.* I could fix both chemical potentials and let both compositions float, and then extract results corresponding to the compositions of interest, but that seems like it could be pretty expensive, especially because I have little sense of the right chemical potential range for the fixed-composition species.

*In fact the situation is that both species can formally be said to inhabit the same sublattice, but one occupies a set of distorted sites that change the cluster orbits. To address this we created a new sublattice consisting of the distorted sites. Experiments find that there is a site-exclusion effect, in which there is no simultaneous occupancy of both the regular site and any of the associated distorted sites. We think that we can enforce this by including a term in the CE that effectively excludes such configurations, but we will have to insert it post facto since there are no such clusters appearing in our training sets.(**) Alternatively we could exclude such clusters from the MC, if that were a possibility.

**Edit: I may have been wrong about that last part. We do seem to have a term that represents that interaction.

**Comment by dmebane** [comment]
Hi -- update: I am using a workaround which is to sweep chemical potentials of the composition I want to fix, then fitting a function to the results and fixing that function at the composition I want. This seems to work at the cost of a lot of extra computation.

**Reply by lbluque** [reply]
Hello @dmebane,

Your method works, but as you mention seem very computationally inefficient. 

For the case of sublattices that effectively do not share any species we can implement a "hybrid" MC step proposal class, that can allow different types of steps per sublattice---for example, grand-canonical steps for one and canonical steps for another. Would this address your situation?

**Reply by dmebane** [reply]
Hi @lbluque -- absolutely! This is precisely what we need.

**Reply by lbluque** [reply]
That should be straightforward to implement. I'll try to have a stab at it hopefully by tomorrow.

**Reply by qchempku2017** [reply]
Hi @lbluque , sorry for delayed response. I've recently returned to China and struggled for a long while for getting access to github. I suggest we implement a simple hybrid MCusher class (maybe called CompositeMcUsher) that takes as input multiple MCUshers. At proposal, we can implement one of the following, or implement both while adding an option for the users to select:

1(Multi MC step), At each MC step, each Usher in the composite should propose a step, and the final MC step should be a combination of proposals from all the Ushers. This requires the sublattices for each Usher not to overlap, otherwise equal a-priori criteria will be broken. We can add sublattice overlap as a check at class initialization. This implementation could benefit from having multiple sublattices done at once, but multi-swaps can also increase the probability of unfavorable configurations. 

2(Single MC step), Choose one usher to execute at a time when proposing a step at equal probability or some user-input probability.

As you go, keep in mind that the importance pre-factor (maybe called compute_log_priori in the usher class) might need to be treated carefully.

**Reply by dmebane** [reply]
Thanks to both of you! This will be perfect for comparing with experiment in materials with mixed occupancy on multiple sublattices, where only one has an experimentally fixed chemical potential -- e.g. many Li+, O2- and H+ conductors 🙏

**Reply by dmebane** [reply]
BTW we are getting useful results -- great application!

**Reply by dmebane** [reply]
Hi, was just wondering if there are any updates . . . much appreciated!

**Reply by lbluque** [reply]
Hi @dmebane,

Apologies for the long period of inactivity and thanks for the ping - this task had slipped my mind (both myself and @qchempku2017 started new positions)

I'll aim to draft a PR this week or next!

**Reply by dmebane** [reply]
ok thanks so much!

**Reply by kamronald** [reply, marked as answer]
Hi @dmebane 
Just looked into this, sorry for the delay. We have a CompositeUsher already implemented, which in principle can solve your problem. However, it is currently not very user-friendly, as you have to manually create a list of MCUshers to pass into the Sampler constructor. Also, I don't think it yet works for the "table-flip" step type.
Here is example code to make a Sampler with a CompositeUsher, assuming you have an ensemble made already:
```
from smol.moca import Ensemble, Sampler
from smol.moca.kernel.mcusher import mcusher_factory, available_step_types

sublattices = ensemble.processor.get_sublattices()
usher1 = mcusher_factory(usher_type='flip', sublattices=[sublattices[0]])
usher2 = mcusher_factory('swap', [sublattices[1]])
usher3 = mcusher_factory('swap', [sublattices[2]])
usher_l = [usher1, usher2, usher3]
sampler = Sampler.from_ensemble(ensemble, step_type='composite', mcushers=usher_l, temperature=1000)
```

I have tried making one of the ushers with the 'table-flip' step type using
`usher1 = mcusher_factory('table-flip', sublattices[0])`
, but am running into an error which I am trying to work out. In the meantime, if the "flip" and "swap" MCUsher types work for you, then you can go ahead to use that.

**Reply by dmebane** [reply]
Hi Ronald -- thanks so much! Just in time as I have received some new CE results and need to run some MC. So to be clear, the 'flip' sublattice will run with fixed chemical potential, while the 'swap' sublattices will run with fixed concentration?

**Reply by dmebane** [reply]
Works! Thanks so much for the assistance, everyone.

**Reply by lbluque** [reply]
Glad to hear it!

Thanks @kamronald for looking into this!

Please, let us know if you have any issues or suggestions in improving the user interface @dmebane

**Reply by dmebane** [reply]
Hi @lbluque thanks so much. Indeed there is a kind of coda here -- the method works in the manner outlined by @kamronald, but we do have to fix the concentration on the sublattice that is using the 'swap' usher. We do that using the Composition subclass from pymatgen:

`from pymatgen.core.composition import Composition`
`sublattices = ensemble.processor.get_sublattices()`
`compositions = [sublattice.composition for sublattice in ensemble.sublattices]`
`compositions[0] = Composition('Li0.291015625')`
`compositions[1] = Composition('La0.5703125')`

which works fine until we want to add an Ewald term. For that we have to name the species with formal charges -- 'Li+', etc. -- but Composition won't accept that, returning an error something like 'you can't use + in a concentration'. Any workaround there?

**Reply by kamronald** [reply]
Hi @dmebane, I think as long as the initial occupancy vector that you pass into the sampler contains the compositions of each sublattice that you desire, you won't need to create these composition objects. By definition, the "swap" steps will maintain the original composition of the sublattice with the swap usher.

**Reply by dmebane** [reply]
Hi @kamronald -- I figured as much regarding the swap steps, and indeed that works great. But the need to use Composition objects arose precisely in calling 'generate_random_ordered_occupancy', which is the only method of defining an initial occupancy vector I'm familiar with.

`init_occu = generate_random_ordered_occupancy(ensemble.processor, composition=compositions)`

**Reply by kamronald** [reply]
Got it, that makes sense. There is a different way to make composition objects, by first making a dictionary with the keys corresponding to a species string (oxidation states are allowed), and the values being the respective occupation fractions. So in your first composition you could do the following:
comp0_dict = {"Li+": 0.291015625}
comp0 = Composition(comp0_dict)
Does this solve the issue?

**Reply by dmebane** [reply]
'generate_random_ordered_accuracy' throws

ValueError: species are present in composition that are not in sublattices.

`comp_dict = {"Li+": 0.291015625, "La3+": 0.5703125, "Ti4+": 1, "O2-": 1}`
`compositions = Composition(comp_dict)`

print(sublattices)
[Sublattice(site_space=Li+0.125 vacA0+0.875, sites=array([   0,    1,    2, ..., 2045, 2046, 2047]), active_sites=array([   0,    1,    2, ..., 2045, 2046, 2047]), encoding=array([0, 1])), Sublattice(site_space=La3+0.5 vacA0+0.5, sites=array([2048, 2049, 2050, 2051, 2052, 2053, 2054, 2055, 2056, 2057, 2058,
       2059, 2060, 2061, 2062, 2063, 2064, 2065, 2066, 2067, 2068, 2069,
       2070, 2071, 2072, 2073, 2074, 2075, 2076, 2077, 2078, 2079, 2080,
       2081, 2082, 2083, 2084, 2085, 2086, 2087, 2088, 2089, 2090, 2091,
       2092, 2093, 2094, 2095, 2096, 2097, 2098, 2099, 2100, 2101, 2102,
       2103, 2104, 2105, 2106, 2107, 2108, 2109, 2110, 2111, 2112, 2113,
       2114, 2115, 2116, 2117, 2118, 2119, 2120, 2121, 2122, 2123, 2124,
       2125, 2126, 2127, 2128, 2129, 2130, 2131, 2132, 2133, 2134, 2135,
       2136, 2137, 2138, 2139, 2140, 2141, 2142, 2143, 2144, 2145, 2146,
       2147, 2148, 2149, 2150, 2151, 2152, 2153, 2154, 2155, 2156, 2157,
       2158, 2159, 2160, 2161, 2162, 2163, 2164, 2165, 2166, 2167, 2168,
       2169, 2170, 2171, 2172, 2173, 2174, 2175, 2176, 2177, 2178, 2179,
       2180, 2181, 2182, 2183, 2184, 2185, 2186, 2187, 2188, 2189, 2190,
       2191, 2192, 2193, 2194, 2195, 2196, 2197, 2198, 2199, 2200, 2201,
       2202, 2203, 2204, 2205, 2206, 2207, 2208, 2209, 2210, 2211, 2212,
       2213, 2214, 2215, 2216, 2217, 2218, 2219, 2220, 2221, 2222, 2223,
       2224, 2225, 2226, 2227, 2228, 2229, 2230, 2231, 2232, 2233, 2234,
       2235, 2236, 2237, 2238, 2239, 2240, 2241, 2242, 2243, 2244, 2245,
       2246, 2247, 2248, 2249, 2250, 2251, 2252, 2253, 2254, 2255, 2256,
       2257, 2258, 2259, 2260, 2261, 2262, 2263, 2264, 2265, 2266, 2267,
       2268, 2269, 2270, 2271, 2272, 2273, 2274, 2275, 2276, 2277, 2278,
       2279, 2280, 2281, 2282, 2283, 2284, 2285, 2286, 2287, 2288, 2289,
       2290, 2291, 2292, 2293, 2294, 2295, 2296, 2297, 2298, 2299, 2300,
       2301, 2302, 2303, 2304, 2305, 2306, 2307, 2308, 2309, 2310, 2311,
       2312, 2313, 2314, 2315, 2316, 2317, 2318, 2319, 2320, 2321, 2322,
       2323, 2324, 2325, 2326, 2327, 2328, 2329, 2330, 2331, 2332, 2333,
       2334, 2335, 2336, 2337, 2338, 2339, 2340, 2341, 2342, 2343, 2344,
       2345, 2346, 2347, 2348, 2349, 2350, 2351, 2352, 2353, 2354, 2355,
       2356, 2357, 2358, 2359, 2360, 2361, 2362, 2363, 2364, 2365, 2366,
       2367, 2368, 2369, 2370, 2371, 2372, 2373, 2374, 2375, 2376, 2377,
       2378, 2379, 2380, 2381, 2382, 2383, 2384, 2385, 2386, 2387, 2388,
       2389, 2390, 2391, 2392, 2393, 2394, 2395, 2396, 2397, 2398, 2399,
       2400, 2401, 2402, 2403, 2404, 2405, 2406, 2407, 2408, 2409, 2410,
       2411, 2412, 2413, 2414, 2415, 2416, 2417, 2418, 2419, 2420, 2421,
       2422, 2423, 2424, 2425, 2426, 2427, 2428, 2429, 2430, 2431, 2432,
       2433, 2434, 2435, 2436, 2437, 2438, 2439, 2440, 2441, 2442, 2443,
       2444, 2445, 2446, 2447, 2448, 2449, 2450, 2451, 2452, 2453, 2454,
       2455, 2456, 2457, 2458, 2459, 2460, 2461, 2462, 2463, 2464, 2465,
       2466, 2467, 2468, 2469, 2470, 2471, 2472, 2473, 2474, 2475, 2476,
       2477, 2478, 2479, 2480, 2481, 2482, 2483, 2484, 2485, 2486, 2487,
       2488, 2489, 2490, 2491, 2492, 2493, 2494, 2495, 2496, 2497, 2498,
       2499, 2500, 2501, 2502, 2503, 2504, 2505, 2506, 2507, 2508, 2509,
       2510, 2511, 2512, 2513, 2514, 2515, 2516, 2517, 2518, 2519, 2520,
       2521, 2522, 2523, 2524, 2525, 2526, 2527, 2528, 2529, 2530, 2531,
       2532, 2533, 2534, 2535, 2536, 2537, 2538, 2539, 2540, 2541, 2542,
       2543, 2544, 2545, 2546, 2547, 2548, 2549, 2550, 2551, 2552, 2553,
       2554, 2555, 2556, 2557, 2558, 2559]), active_sites=array([2048, 2049, 2050, 2051, 2052, 2053, 2054, 2055, 2056, 2057, 2058,
       2059, 2060, 2061, 2062, 2063, 2064, 2065, 2066, 2067, 2068, 2069,
       2070, 2071, 2072, 2073, 2074, 2075, 2076, 2077, 2078, 2079, 2080,
       2081, 2082, 2083, 2084, 2085, 2086, 2087, 2088, 2089, 2090, 2091,
       2092, 2093, 2094, 2095, 2096, 2097, 2098, 2099, 2100, 2101, 2102,
       2103, 2104, 2105, 2106, 2107, 2108, 2109, 2110, 2111, 2112, 2113,
       2114, 2115, 2116, 2117, 2118, 2119, 2120, 2121, 2122, 2123, 2124,
       2125, 2126, 2127, 2128, 2129, 2130, 2131, 2132, 2133, 2134, 2135,
       2136, 2137, 2138, 2139, 2140, 2141, 2142, 2143, 2144, 2145, 2146,
       2147, 2148, 2149, 2150, 2151, 2152, 2153, 2154, 2155, 2156, 2157,
       2158, 2159, 2160, 2161, 2162, 2163, 2164, 2165, 2166, 2167, 2168,
       2169, 2170, 2171, 2172, 2173, 2174, 2175, 2176, 2177, 2178, 2179,
       2180, 2181, 2182, 2183, 2184, 2185, 2186, 2187, 2188, 2189, 2190,
       2191, 2192, 2193, 2194, 2195, 2196, 2197, 2198, 2199, 2200, 2201,
       2202, 2203, 2204, 2205, 2206, 2207, 2208, 2209, 2210, 2211, 2212,
       2213, 2214, 2215, 2216, 2217, 2218, 2219, 2220, 2221, 2222, 2223,
       2224, 2225, 2226, 2227, 2228, 2229, 2230, 2231, 2232, 2233, 2234,
       2235, 2236, 2237, 2238, 2239, 2240, 2241, 2242, 2243, 2244, 2245,
       2246, 2247, 2248, 2249, 2250, 2251, 2252, 2253, 2254, 2255, 2256,
       2257, 2258, 2259, 2260, 2261, 2262, 2263, 2264, 2265, 2266, 2267,
       2268, 2269, 2270, 2271, 2272, 2273, 2274, 2275, 2276, 2277, 2278,
       2279, 2280, 2281, 2282, 2283, 2284, 2285, 2286, 2287, 2288, 2289,
       2290, 2291, 2292, 2293, 2294, 2295, 2296, 2297, 2298, 2299, 2300,
       2301, 2302, 2303, 2304, 2305, 2306, 2307, 2308, 2309, 2310, 2311,
       2312, 2313, 2314, 2315, 2316, 2317, 2318, 2319, 2320, 2321, 2322,
       2323, 2324, 2325, 2326, 2327, 2328, 2329, 2330, 2331, 2332, 2333,
       2334, 2335, 2336, 2337, 2338, 2339, 2340, 2341, 2342, 2343, 2344,
       2345, 2346, 2347, 2348, 2349, 2350, 2351, 2352, 2353, 2354, 2355,
       2356, 2357, 2358, 2359, 2360, 2361, 2362, 2363, 2364, 2365, 2366,
       2367, 2368, 2369, 2370, 2371, 2372, 2373, 2374, 2375, 2376, 2377,
       2378, 2379, 2380, 2381, 2382, 2383, 2384, 2385, 2386, 2387, 2388,
       2389, 2390, 2391, 2392, 2393, 2394, 2395, 2396, 2397, 2398, 2399,
       2400, 2401, 2402, 2403, 2404, 2405, 2406, 2407, 2408, 2409, 2410,
       2411, 2412, 2413, 2414, 2415, 2416, 2417, 2418, 2419, 2420, 2421,
       2422, 2423, 2424, 2425, 2426, 2427, 2428, 2429, 2430, 2431, 2432,
       2433, 2434, 2435, 2436, 2437, 2438, 2439, 2440, 2441, 2442, 2443,
       2444, 2445, 2446, 2447, 2448, 2449, 2450, 2451, 2452, 2453, 2454,
       2455, 2456, 2457, 2458, 2459, 2460, 2461, 2462, 2463, 2464, 2465,
       2466, 2467, 2468, 2469, 2470, 2471, 2472, 2473, 2474, 2475, 2476,
       2477, 2478, 2479, 2480, 2481, 2482, 2483, 2484, 2485, 2486, 2487,
       2488, 2489, 2490, 2491, 2492, 2493, 2494, 2495, 2496, 2497, 2498,
       2499, 2500, 2501, 2502, 2503, 2504, 2505, 2506, 2507, 2508, 2509,
       2510, 2511, 2512, 2513, 2514, 2515, 2516, 2517, 2518, 2519, 2520,
       2521, 2522, 2523, 2524, 2525, 2526, 2527, 2528, 2529, 2530, 2531,
       2532, 2533, 2534, 2535, 2536, 2537, 2538, 2539, 2540, 2541, 2542,
       2543, 2544, 2545, 2546, 2547, 2548, 2549, 2550, 2551, 2552, 2553,
       2554, 2555, 2556, 2557, 2558, 2559]), encoding=array([0, 1])), Sublattice(site_space=Ti4+1 , sites=array([2560, 2561, 2562, 2563, 2564, 2565, 2566, 2567, 2568, 2569, 2570,
       2571, 2572, 2573, 2574, 2575, 2576, 2577, 2578, 2579, 2580, 2581,
       2582, 2583, 2584, 2585, 2586, 2587, 2588, 2589, 2590, 2591, 2592,
       2593, 2594, 2595, 2596, 2597, 2598, 2599, 2600, 2601, 2602, 2603,
       2604, 2605, 2606, 2607, 2608, 2609, 2610, 2611, 2612, 2613, 2614,
       2615, 2616, 2617, 2618, 2619, 2620, 2621, 2622, 2623, 2624, 2625,
       2626, 2627, 2628, 2629, 2630, 2631, 2632, 2633, 2634, 2635, 2636,
       2637, 2638, 2639, 2640, 2641, 2642, 2643, 2644, 2645, 2646, 2647,
       2648, 2649, 2650, 2651, 2652, 2653, 2654, 2655, 2656, 2657, 2658,
       2659, 2660, 2661, 2662, 2663, 2664, 2665, 2666, 2667, 2668, 2669,
       2670, 2671, 2672, 2673, 2674, 2675, 2676, 2677, 2678, 2679, 2680,
       2681, 2682, 2683, 2684, 2685, 2686, 2687, 2688, 2689, 2690, 2691,
       2692, 2693, 2694, 2695, 2696, 2697, 2698, 2699, 2700, 2701, 2702,
       2703, 2704, 2705, 2706, 2707, 2708, 2709, 2710, 2711, 2712, 2713,
       2714, 2715, 2716, 2717, 2718, 2719, 2720, 2721, 2722, 2723, 2724,
       2725, 2726, 2727, 2728, 2729, 2730, 2731, 2732, 2733, 2734, 2735,
       2736, 2737, 2738, 2739, 2740, 2741, 2742, 2743, 2744, 2745, 2746,
       2747, 2748, 2749, 2750, 2751, 2752, 2753, 2754, 2755, 2756, 2757,
       2758, 2759, 2760, 2761, 2762, 2763, 2764, 2765, 2766, 2767, 2768,
       2769, 2770, 2771, 2772, 2773, 2774, 2775, 2776, 2777, 2778, 2779,
       2780, 2781, 2782, 2783, 2784, 2785, 2786, 2787, 2788, 2789, 2790,
       2791, 2792, 2793, 2794, 2795, 2796, 2797, 2798, 2799, 2800, 2801,
       2802, 2803, 2804, 2805, 2806, 2807, 2808, 2809, 2810, 2811, 2812,
       2813, 2814, 2815, 2816, 2817, 2818, 2819, 2820, 2821, 2822, 2823,
       2824, 2825, 2826, 2827, 2828, 2829, 2830, 2831, 2832, 2833, 2834,
       2835, 2836, 2837, 2838, 2839, 2840, 2841, 2842, 2843, 2844, 2845,
       2846, 2847, 2848, 2849, 2850, 2851, 2852, 2853, 2854, 2855, 2856,
       2857, 2858, 2859, 2860, 2861, 2862, 2863, 2864, 2865, 2866, 2867,
       2868, 2869, 2870, 2871, 2872, 2873, 2874, 2875, 2876, 2877, 2878,
       2879, 2880, 2881, 2882, 2883, 2884, 2885, 2886, 2887, 2888, 2889,
       2890, 2891, 2892, 2893, 2894, 2895, 2896, 2897, 2898, 2899, 2900,
       2901, 2902, 2903, 2904, 2905, 2906, 2907, 2908, 2909, 2910, 2911,
       2912, 2913, 2914, 2915, 2916, 2917, 2918, 2919, 2920, 2921, 2922,
       2923, 2924, 2925, 2926, 2927, 2928, 2929, 2930, 2931, 2932, 2933,
       2934, 2935, 2936, 2937, 2938, 2939, 2940, 2941, 2942, 2943, 2944,
       2945, 2946, 2947, 2948, 2949, 2950, 2951, 2952, 2953, 2954, 2955,
       2956, 2957, 2958, 2959, 2960, 2961, 2962, 2963, 2964, 2965, 2966,
       2967, 2968, 2969, 2970, 2971, 2972, 2973, 2974, 2975, 2976, 2977,
       2978, 2979, 2980, 2981, 2982, 2983, 2984, 2985, 2986, 2987, 2988,
       2989, 2990, 2991, 2992, 2993, 2994, 2995, 2996, 2997, 2998, 2999,
       3000, 3001, 3002, 3003, 3004, 3005, 3006, 3007, 3008, 3009, 3010,
       3011, 3012, 3013, 3014, 3015, 3016, 3017, 3018, 3019, 3020, 3021,
       3022, 3023, 3024, 3025, 3026, 3027, 3028, 3029, 3030, 3031, 3032,
       3033, 3034, 3035, 3036, 3037, 3038, 3039, 3040, 3041, 3042, 3043,
       3044, 3045, 3046, 3047, 3048, 3049, 3050, 3051, 3052, 3053, 3054,
       3055, 3056, 3057, 3058, 3059, 3060, 3061, 3062, 3063, 3064, 3065,
       3066, 3067, 3068, 3069, 3070, 3071]), active_sites=array([], dtype=float64), encoding=array([0])), Sublattice(site_space=O2-1 , sites=array([3072, 3073, 3074, ..., 4605, 4606, 4607]), active_sites=array([], dtype=float64), encoding=array([0]))]

**Reply by kamronald** [reply]
I think the issue is that you created one composition object, but what the `generate_random_ordered_accuracy` function needs is a list of compositions, one for each sublattice. To create that list, you could do something like this:

```
from pymatgen.core import Composition
from smol.cofe.space import Vacancy
comp0 = Composition({"Li+": 0.125, Vacancy(): 0.875})
comp1 = Composition({"La3+": 0.3, Vacancy(): 0.7})
comp2 = Composition("Ti4+")
comp3 = Composition("O2-")
compositions = [comp0, comp1, comp2, comp3]
```

**Reply by kamronald** [reply]
Another way to do this is to use Pymatgen's OrderDisorderedStructureTransformation. You would just need to ensure that the disordered structure you pass in has a correct supercell size for the composition you want (ie the composition leads to an integer number of each species for that supercell size). Then it can generate ordered structures of lowest Ewald energy. If you use this, I would recommend algo=2 for best speed. https://pymatgen.org/pymatgen.transformations.html

**Reply by dmebane** [reply]
yes that does it thanks!
